ldif.go

```diff
@@ -0,0 +1,256 @@
+						lines = append(lines, line)
+					}
+					line = nextLine
+					continue
+				}
+			}
+		default:
+			return l.newError(err.Error())
+		}
+	}
+	if len(lines) != 0 {
+		entry, perr := l.parseEntry(lines)
+		if perr != nil {
+			return l.newError(perr.Error())
+		}
+		l.Entries = append(l.Entries, entry)
+	}
+	return nil
+}
+
+func (l *LDIF) parseEntry(lines [][]byte) (entry *Entry, err error) {
+	// for i, line := range lines {
+	//     fmt.Fprintf(os.Stderr, "% 2d %s\n", i, line)
+	// }
+	if l.version == 0 && bytes.HasPrefix(lines[0], []byte("version:")) {
+		line := bytes.TrimLeft(lines[0][8:], SPACES)
+		if version, err := strconv.Atoi(string(line)); err != nil {
+			return nil, err
+		} else {
+			if version != 1 {
+				return nil, errors.New("Invalid version spec " + string(line))
+			}
+			l.version = 1
+			lines = lines[1:]
+		}
+	}
+
+	if !bytes.HasPrefix(lines[0], []byte("dn:")) {
+		return nil, errors.New("Missing dn:")
+	}
+	_, val, err := l.parseLine(lines[0])
+	if err != nil {
+		return nil, err
+	}
+	dn := val
+
+	lines = lines[1:]
+	if bytes.HasPrefix(lines[0], []byte("changetype:")) {
+		_, val, err := l.parseLine(lines[0])
+		if err != nil {
+			return nil, err
+		}
+		l.changeType = val
+		lines = lines[1:]
+	}
+	if l.changeType != "" {
+		return nil, errors.New("change records not supported")
+	}
+
+	attrs := make(map[string][]string)
+	for i := 0; i < len(lines); i++ {
+		attr, val, err := l.parseLine(lines[i])
+		if err != nil {
+			if !l.RelaxedParser {
+				return nil, err
+			} else {
+				continue
+			}
+		}
+		if _, ok := attrs[attr]; ok {
+			attrs[attr] = append(attrs[attr], string(val))
+		} else {
+			attrs[attr] = []string{string(val)}
+		}
+	}
+	return NewEntry(dn, attrs), nil
+}
+
+func (l *LDIF) parseLine(line []byte) (attr, val string, err error) {
+	off := 0
+	for len(line) > off && line[off] != ':' {
+		off++
+		if off >= len(line) {
+			return
+		}
+	}
+	if off == len(line) {
+		err = errors.New("Missing : in line")
+		return
+	}
+	if off > len(line)-2 {
+		err = errors.New("empty value")
+		return
+	}
+
+	attr = string(line[0:off])
+	if err = validAttr(attr); err != nil {
+		attr = ""
+		val = ""
+		return
+	}
+
+	switch line[off+1] {
+	case ':':
+		var n int
+		value := bytes.TrimLeft(line[off+2:], SPACES)
+		// fmt.Fprintf(os.Stderr, "LINE=%s\nVALUE=%s\n", line, value)
+		dec := make([]byte, base64.StdEncoding.DecodedLen(len(value)))
+		n, err = base64.StdEncoding.Decode(dec, value)
+		if err != nil {
+			return
+		}
+		val = string(dec[:n])
+	case '<': // FIXME missing return for *net.URL type
+		val = string(bytes.TrimLeft(line[off+2:], SPACES))
+	default:
+		val = string(bytes.TrimLeft(line[off+2:], SPACES))
+	}
+
+	return
+}
+
+func validOID(oid string) error {
+	lastDot := true
+	for _, c := range oid {
+		switch {
+		case c == '.' && lastDot:
+			return errors.New("OID with at least 2 consecutive dots")
+		case c == '.':
+			lastDot = true
+		case c >= 0x30 && c <= 0x39:
```

Comment:
does comparing to runes work (`c >= '0' && c <= '9'`)? if so, it would be more readable